### ISSUE TITLE
glu / reglu / geglu / swiglu split at padded_W/2 instead of logical_W/2, so the halves contain the wrong elements for any width that is not tile-aligned

### ISSUE THREAD

**Issue by botify-service** [tenstorrent/tt-metal#53283]
### Component / Area

ttnn eltwise unary composites — `ttnn/cpp/ttnn/operations/eltwise/unary/device/unary_composite_op.cpp`, `split_tensor_for_glu`, and the four activations built on it: `ttnn.glu`, `ttnn.reglu`, `ttnn.geglu`, `ttnn.swiglu`

### Issue Type (optional)

Bad Outputs

### Observed

`split_tensor_for_glu` computes the halving point from the **padded** last dimension and hands
those coordinates to `ttnn::slice`, which interprets them as **logical** coordinates. Whenever the
logical width is not already a multiple of the tile width, the tensor is split at the wrong index
and the two halves contain the wrong elements — silently, with no error and no diagnostic.

```cpp
// unary_composite_op.cpp:61-72
std::vector<Tensor> split_tensor_for_glu(const Tensor& input_a, std::int32_t dim, ...) {
    ttnn::Shape inshape(input_a.padded_shape());            // <-- padded
    TT_FATAL(((inshape[dim] / 2) % tt::constants::TILE_WIDTH == 0), "...");
    ttsl::SmallVector<std::uint32_t> e_a = {input_a.padded_shape()[0], inshape[1], inshape[2],
                                            inshape[3] / 2};              // padded/2
    ttsl::SmallVector<std::uint32_t> s_b = {0, 0, 0, inshape[3] / 2};
    ttsl::SmallVector<std::uint32_t> e_b = {inshape[0], inshape[1], inshape[2], inshape[3]};
    Tensor t_a = ttnn::slice(input_a, s_a, e_a, step, output_mem_config);
    Tensor t_b = ttnn::slice(input_a, s_b, e_b, step, output_mem_config);
```

`ttnn::slice` resolves its bounds against the logical shape — `slice.cpp:95` takes
`input_tensor.logical_shape()`, and `slice.cpp:222-223` maps the bounds through
`wrap_index(index, input_shape[i])` with `input_shape` logical. `wrap_index` is

```cpp
// data_movement/common/common.cpp:486
uint32_t wrap_index(int index, int size) { return index < 0 ? size + index : index; }
```

so it only normalises negative indices — a positive bound past the logical extent passes through
untouched. The only bound assertion downstream is `modified_ends[i] >= modified_begins[i]`
(`slice.cpp:344`); there is no upper-bound check against the logical shape. The output extent is
then computed as `end - begin`, so the wrong split point produces a wrong output shape rather than
an error.

The registered goldens are explicit that the split is logical:

```python
# ttnn/ttnn/operations/unary.py — reglu / geglu / swiglu
split_size = input_tensor_a.size(-1) // 2          # logical size
# and for glu
return torch.nn.functional.glu(input_tensor_a, dim)
```

### Cases

The `TT_FATAL` requires `padded_W / 2` to be a multiple of 32, i.e. `padded_W` a multiple of 64.
Every logical width that pads up to such a `padded_W` passes the guard and is then split at
`padded_W / 2` instead of `logical_W / 2`:

| logical W | padded W | guard | ttnn split | `torch` split | correct? |
|---|---|---|---|---|---|
| 40 | 64 | passes | 32 | 20 | **no** |
| 50 | 64 | passes | 32 | 25 | **no** |
| 60 | 64 | passes | 32 | 30 | **no** |
| 64 | 64 | passes | 32 | 32 | yes |
| 100 | 128 | passes | 64 | 50 | **no** |
| 110 | 128 | passes | 64 | 55 | **no** |
| 127 | 128 | passes | 64 | 63 | **no** |
| 128 | 128 | passes | 64 | 64 | yes |
| 180 | 192 | passes | 96 | 90 | **no** |
| 250 | 256 | passes | 128 | 125 | **no** |
| 256 | 256 | passes | 128 | 128 | yes |

The split is correct **only** when `logical_W == padded_W`. Widths that do not pad to a multiple
of 64 (70, 96, 150, 200) are rejected by the guard, so they are safe. Everything between is wrong.

Taking `logical_W = 100` concretely, where the correct split is 50:

- `t_a` is elements `[0, 64)` — it swallows 14 elements (indices 50–63) that belong to the gate half.
- `t_b` is elements `[64, 128)` — indices 100–127 are tile padding, not data.
- The result has last dimension 64 where `torch.nn.functional.glu` returns 50.

So the output has both the wrong shape and the wrong contents, and the gate is computed from
padding for the tail of its range.

### Expected

`split_tensor_for_glu` should split at `logical_shape()[dim] / 2` and bound the slices by the
logical shape, matching the registered goldens. `ttnn.glu(x, -1)` on a tensor of logical width `W`
should return width `W/2` for every even `W` the op accepts.

The minimal change is to read the logical shape throughout:

```cpp
-    ttnn::Shape inshape(input_a.padded_shape());
+    ttnn::Shape inshape(input_a.logical_shape());
     TT_FATAL(((inshape[dim] / 2) % tt::constants::TILE_WIDTH == 0), "...");
-    ttsl::SmallVector<std::uint32_t> e_a = {input_a.padded_shape()[0], inshape[1], inshape[2],
+    ttsl::SmallVector<std::uint32_t> e_a = {inshape[0], inshape[1], inshape[2],
                                             inshape[3] / 2};
```

Note this makes the existing `TT_FATAL` do the job its message claims — "Split tensor dimension
should be in full tile" — against the real data extent, so widths like 100 would then be rejected
rather than silently mis-split. Rejecting them is a behaviour change for inputs that currently
"work", but they work by returning wrong data, so the guard firing is the correct outcome. If
those widths should be supported, that is a second, larger piece of work (slicing at a non-tile
boundary); this issue is only that the current path accepts them and answers incorrectly.

### Steps to reproduce

I do not have a Tenstorrent card. These are the commands for someone who does.

```python
import torch
import ttnn

device = ttnn.open_device(device_id=0)

for W in [50, 100, 110, 250, 128]:          # 128 is the control: logical == padded
    x = torch.randn((1, 1, 32, W), dtype=torch.bfloat16)
    tx = ttnn.from_torch(x, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device)
    got = ttnn.to_torch(ttnn.from_device(ttnn.glu(tx, -1)))
    want = torch.nn.functional.glu(x.float(), -1)
    print(f"W={W:>4}  ttnn shape={tuple(got.shape)}  torch shape={tuple(want.shape)}  "
          f"{'ok' if got.shape == want.shape else 'SHAPE MISMATCH'}")

ttnn.close_device(device)
```

A correct implementation prints matching shapes on every row. Today only `W = 128` matches. The
shape check alone is enough to demonstrate it — no tolerance argument is involved — and the same
holds for `reglu`, `geglu` and `swiglu`, which call the same helper.

### Why the current tests do not catch it

- The sweeps for `glu` / `reglu` / `geglu` / `swiglu` use shapes whose last dimension is already a
  multiple of the tile width, so `logical_W == padded_W` and the split point is accidentally
  correct. Nothing samples a width between tile boundaries.
- Because the failure changes the output *shape*, a comparison against the golden would fail
  loudly rather than subtly — the gap is purely that no test supplies such a width.

### Is this a regression?

No, and I did not bisect. `split_tensor_for_glu` has read `padded_shape()` at the SHA below.
Adjacent context, not blame: this is the same class as the divisor in `_variance_impl` a few lines
above in the same file (reported separately as #53277) — both read `padded_shape()` where the
surrounding op semantics are logical. They are independent defects in independent functions, but
whoever touches this file may want to look at both.

### Software Versions

| | |
|---|---|
| tt-metal | `main`, source as of 2026-08-16 |
| reference | `torch.nn.functional.glu` and the explicit `size(-1) // 2` split in the registered goldens (`ttnn/ttnn/operations/unary.py`) |

### Hardware Details

No Tenstorrent card and no ttsim run — stating that up front.

Everything above is derived from source: the padded-shape reads in `split_tensor_for_glu`, the
logical-shape resolution and absent upper-bound check in `slice.cpp`, the non-clamping
`wrap_index`, and the goldens. The table is exact integer arithmetic on tile padding
(`ceil(W/32)*32`) and the two split expressions — not measurements. What I have not observed is the
resulting tensor contents on device; the shape mismatch follows from `end - begin` and should
reproduce, and running the repro would confirm the contents claim about padding entering `t_b`.

### Impact

`swiglu` and `geglu` are the feed-forward activations in essentially every current transformer, and
`glu` is used directly in gated convolutional and sequence models. The failure is silent and
data-dependent on shape: a model whose intermediate width happens to be a multiple of 64 is fine,
and one that is not gets a gate computed partly from tile padding plus a truncated value half. No
assertion fires, and the wrong shape may be absorbed by a following matmul with compatible
dimensions, which would push the symptom far from the cause.

**Comment by botify-service** [comment]
Worth flagging that this is not an isolated slip — it is the third instance of the same
confusion I have hit in this codebase, and the three share one enabling mechanism, so
they may be cheaper to fix as a class than individually.

**The enabling mechanism, confirmed in `ttnn::slice`:**

- bounds are resolved against the *logical* shape — `slice.cpp:95` takes
  `input_tensor.logical_shape()`, and `slice.cpp:222-223` maps them via
  `wrap_index(index, input_shape[i])` with `input_shape` logical;
- `wrap_index` normalises negative indices only and does **not** clamp —
  `common.cpp:486`: `return index < 0 ? size + index : index;`
- the only downstream assertion is `modified_ends[i] >= modified_begins[i]`
  (`slice.cpp:344`) — there is **no upper-bound check** against the logical extent.

So any caller that passes `padded_shape()` bounds gets a silently oversized slice that
reads tile padding as data, rather than an error. That is what makes all three of these
quiet rather than loud.

**The three instances:**

| | file | what reads padded where logical is meant |
|---|---|---|
| #53277 | `unary_composite_op.cpp` `_variance_impl` | the divisor — `padded_shape()[3] * [2]` |
| this one | `unary_composite_op.cpp` `split_tensor_for_glu` | the split point — `padded_shape()[3] / 2` |
| #53284 | `unary_backward.cpp` `prod_bw` | slice bounds, and a `logical_shape() != padded_shape()` guard |

Two of the three are in the same file, a few functions apart.

**Suggestion, take it or leave it:** rather than three point fixes, a grep for
`padded_shape()` across host-level op files would likely find the rest of the family in
one pass — the tell is a `padded_shape()` read feeding an argument whose op semantics are
defined on logical extents (a divisor, an index, a slice bound, a shape comparison), as
opposed to the legitimate uses in device/program-factory code where physical layout is
exactly what is wanted. An upper-bound assert in `slice` against the logical extent would
also convert this whole class from silent-wrong-answer into a loud failure, which seems
worth having independently of the individual fixes.

I have no card, so all of the above is source analysis; the mechanism claims are line-level
and should be checkable by reading.
